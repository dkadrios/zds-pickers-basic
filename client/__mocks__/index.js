export inputControls from './inputControls'
export mappings from './mappings'
export midi from './midi'
export shiftGroups from './shiftGroups'
export shifter from './shifter'
export store from './store'
export user from './user'
export version from './version'
