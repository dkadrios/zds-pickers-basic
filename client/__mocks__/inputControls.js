export default [
  {
    controlId: 0,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 11,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 1,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 1,
    status: 11,
    value: 111,
    active: false,
    lit: false,
  },
  {
    controlId: 2,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
  {
    controlId: 3,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 4,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 5,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
]
