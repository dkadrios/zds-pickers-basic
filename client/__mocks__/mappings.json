{
  "dialogVisible": false,
  "userDialogVisible": false,
  "userImportDialogVisible": false,
  "channels": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    "BFD3",
    null,
    null,
    null,
    null,
    null,
    null
  ],
  "stockMappings": [
    "2box",
    "Addictive Drums",
    "Alesis DM Pro",
    "Alesis DM10",
    "Alesis DM5",
    "BFD2",
    "BFD3",
    "Battery 3",
    "E-MU Protean Acoustic",
    "E-MU Protean Bread n Butter",
    "E-MU Protean Studio",
    "E-MU Protean Synth",
    "EZ Drummer The Classic EZX",
    "EZDrummer",
    "EZDrummer Jazz Kit",
    "EZDrummer Latin Percussion",
    "General MIDI",
    "Infinite Instruments",
    "John Emrich - Articulations 1",
    "Kontakt 2",
    "Ocean Way Drums",
    "Pearl Redbox",
    "Roland HandSonic 15",
    "Roland TD-10",
    "Roland TD-10 Perc 1",
    "Roland TD-10 Perc 2",
    "Roland TD-10 Perc 3",
    "Roland TD-10 Perc 4",
    "Roland TD-12",
    "Roland TD-12 Perc 1 (Latin)",
    "Roland TD-12 Perc 2 (Indian)",
    "Roland TD-12 Perc 3 (African)",
    "Roland TD-12 Perc 4 (Salsa)",
    "Roland TD-12 Perc 5 (SFX)",
    "Roland TD-12 Perc 6-8 (User)",
    "Roland TD-20",
    "Roland TD-20 Perc",
    "Roland TD-3",
    "Roland TD-30",
    "Roland TD-30 Perc 1 (Latin)",
    "Roland TD-30 Perc 2 (Indian)",
    "Roland TD-30 Perc 3 (African)",
    "Roland TD-30 Perc 4 (Salsa)",
    "Roland TD-30 Perc 5 (SFX)",
    "Roland TD-30 Perc 6-8 (User)",
    "Roland TD-5",
    "Roland TD-6",
    "Roland TD-6 Perc 1",
    "Roland TD-6 Perc 10 (Special)",
    "Roland TD-6 Perc 2",
    "Roland TD-6 Perc 3 (Room)",
    "Roland TD-6 Perc 4 (Power)",
    "Roland TD-6 Perc 5 (Elec)",
    "Roland TD-6 Perc 6 (808)",
    "Roland TD-6 Perc 7 (Jazz)",
    "Roland TD-6 Perc 8 (Brush)",
    "Roland TD-6 Perc 9 (Perc)",
    "Roland TD-7",
    "Roland TD-8",
    "Roland TD-8 Perc 1",
    "Roland TD-8 Perc 10 (Special)",
    "Roland TD-8 Perc 2",
    "Roland TD-8 Perc 3 (Room)",
    "Roland TD-8 Perc 4 (Power)",
    "Roland TD-8 Perc 5 (Elec)",
    "Roland TD-8 Perc 6 (808)",
    "Roland TD-8 Perc 7 (Jazz)",
    "Roland TD-8 Perc 8 (Brush)",
    "Roland TD-8 Perc 9 (Perc)",
    "Roland TD-9",
    "Roland TD-9 Perc",
    "Simmons SD1000",
    "Simmons SD5X",
    "Simmons SD7",
    "Simmons SD7PK",
    "Simmons SD9K",
    "Sonic Reality iMap Extended",
    "Sonic Reality iMap Standard",
    "Studio ProFiles",
    "Superior Drummer",
    "Yamaha DTX 700",
    "Yamaha DTXplorer",
    "Zendrum Stompblock",
    "Zendrum TRS",
    "Zendrum TRS 2.0"
  ],
  "userMappings": ["BFD3 Custom ZenEdit", "reaper"]
}
