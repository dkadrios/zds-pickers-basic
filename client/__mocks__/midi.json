{
  "devices": [
    {
      "id": "-1325431060",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "output",
      "version": "",
      "state": "connected",
      "connection": "open"
    },
    {
      "id": "-312145282",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "input",
      "version": "",
      "state": "connected",
      "connection": "open"
    }
  ],
  "listeningDevices": ["-312145282"]
}
