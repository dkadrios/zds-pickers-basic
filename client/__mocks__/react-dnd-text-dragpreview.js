/* eslint-disable import/prefer-default-export, jsx-a11y/alt-text */
import React from 'react'

export const createDragPreview = () => <img src={null} />
