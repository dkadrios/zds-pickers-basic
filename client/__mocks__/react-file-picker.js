/* eslint-disable import/prefer-default-export */
export const FilePicker = () => <div>content</div>
