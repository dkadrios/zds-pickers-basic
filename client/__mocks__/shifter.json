{
  "accessGranted": true,
  "errorMessage": "",
  "errorVisible": false,
  "exportBuffer": [],
  "exportDialogVisible": false,
  "exportFilename": "zds-shifter-backup.txt",
  "found": true,
  "hardwareTestVisible": true,
  "importDialogVisible": false,
  "importInProcess": false,
  "invalidSettingsFile": null,
  "midiActivityLEDMode": 0,
  "midiInActivity": false,
  "midiOutActivity": false,
  "ready": true,
  "resetInProcess": false,
  "responding": true,
  "searchedForShifter": true,
  "serialMidiOutEnabled": true,
  "showResetDialog": false,
  "testInterfaceFound": false,
  "usbMidiOutEnabled": true
}
