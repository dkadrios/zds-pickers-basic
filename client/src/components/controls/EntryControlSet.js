import { Component, createRef } from 'react'
import PropTypes from 'prop-types'
import {
  NotePicker,
  STATUS_CONTROL_CHANGE,
  STATUS_NOTE_OFF,
  STATUS_NOTE_ON,
  STATUS_PITCH_WHEEL,
  StatusPicker,
  ValuePicker,
  statuses,
} from 'zds-pickers'
import { entryControlSet, notePicker, valuePicker } from '../../styles/entryDlg.scss'
import { getMapping, getNoteValue } from '../../reducers/mappings'
import { mappingsShape, midiMessageShape } from '../../core/shapes'
import CCPicker from './CCPicker'
import ChannelPicker from './ChannelPicker'

class EntryControlSet extends Component {
  constructor(props) {
    super(props)
    this.valueControlRef = createRef()
  }

  focusValueControl = () => {
    const { current } = this.valueControlRef
    const control = current.input || current.select
    control.focus()
  }

  render() {
    const {
      changeChannel,
      changeStatus,
      changeValue,
      entry: { status },
      entry,
      groupId,
      isInput,
      mappings: { channels },
      onPressedEnter,
      otherEntry,
    } = this.props

    const statusList = statuses
      // You cannot set input to 'stack' mode
      .filter(({ value }) => !isInput || value !== STATUS_NOTE_OFF)
      // Pitch Wheel is not possible with the Shifter
      .filter(({ value }) => value !== STATUS_PITCH_WHEEL)
      // You MUST use Note On if stacking notes
      .filter(({ value }) => isInput || otherEntry.status === STATUS_NOTE_ON || value !== STATUS_NOTE_OFF)
      .filter(({ value }) => !isInput || otherEntry.status !== STATUS_NOTE_OFF || value === STATUS_NOTE_ON)

    const handleKeyDown = (event) => {
      const { okButtonRef } = this.props
      /* istanbul ignore else */
      if (event.key === 'Tab') {
        event.preventDefault()
      }
      if (event.key === 'Enter' || event.key === 'Tab') {
        onPressedEnter(isInput, event.key === 'Tab', okButtonRef)
      }
    }

    const entryProps = { ...entry }

    if (entryProps.status === STATUS_NOTE_ON) {
      const entryData = getNoteValue(channels, entry.channel, entry.value)
      /* istanbul ignore else */
      if (entryData) {
        entryProps.value = `${entryData.note} ${entryData.name}`
      }
    }

    const pickerProps = {
      onChange: value => changeValue(groupId, isInput, value),
      autoFocus: isInput,
      onKeyDown: handleKeyDown,
      inputRef: this.valueControlRef,
    }

    const renderPicker = () => {
      const mapping = getMapping(channels, entry.channel)

      if (status === STATUS_CONTROL_CHANGE) {
        return (
          <CCPicker
            {...entry}
            {...pickerProps}
          />
        )
      }
      if ((status === STATUS_NOTE_ON || status === STATUS_NOTE_OFF) && mapping) {
        return (
          <NotePicker
            mapping={mapping}
            {...entry}
            {...pickerProps}
            className={notePicker}
          />
        )
      }
      return (
        <ValuePicker
          {...entryProps}
          className={valuePicker}
          ref={this.valueControlRef}
          {...pickerProps}
        />
      )
    }

    return (
      <div className={entryControlSet}>
        <h3>{isInput ? 'Input' : 'Output'}</h3>
        <StatusPicker
          {...entry}
          onChange={value => changeStatus(groupId, isInput, value)}
          statuses={statusList}
        />
        <ChannelPicker
          {...entry}
          onChange={value => changeChannel(groupId, isInput, value)}
        />
        {renderPicker()}
      </div>
    )
  }
}

EntryControlSet.propTypes = {
  isInput: PropTypes.bool,
  entry: midiMessageShape.isRequired,
  groupId: PropTypes.number.isRequired,
  changeStatus: PropTypes.func.isRequired,
  changeValue: PropTypes.func.isRequired,
  changeChannel: PropTypes.func.isRequired,
  onPressedEnter: PropTypes.func.isRequired,
  mappings: mappingsShape.isRequired,
  otherEntry: midiMessageShape.isRequired,
  okButtonRef: PropTypes.object.isRequired,
}

EntryControlSet.defaultProps = {
  isInput: false,
}

export default EntryControlSet
