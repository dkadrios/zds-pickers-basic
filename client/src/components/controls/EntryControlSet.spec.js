import React from 'react'
import set from 'lodash/fp/set'
import { STATUS_CONTROL_CHANGE, STATUS_NOTE_ON } from 'zds-pickers'
import { mappings } from '../../../__mocks__'
import EntryControlSet from './EntryControlSet'

const changeStatus = jest.fn()
const changeValue = jest.fn()
const changeChannel = jest.fn()
const onPressedEnter = jest.fn()

const okButtonRef = {}

const isInput = false
const groupId = 0

const entry = {
  channel: 9,
  status: 13,
  value: 22,
}

const otherEntry = {
  channel: 9,
  status: 23,
  value: 44,
}

const props = {
  isInput,
  entry,
  groupId: 0,
  changeStatus,
  changeValue,
  changeChannel,
  onPressedEnter,
  mappings,
  otherEntry,
  okButtonRef,
}

describe('EntryControlSet tests', () => {
  beforeEach(jest.resetAllMocks)

  it('renders correctly as output', () => {
    shallowExpect(EntryControlSet)(set('isInput', false)(props)).toMatchSnapshot()
  })

  it('renders correctly as input', () => {
    shallowExpect(EntryControlSet)(set('isInput', true)(props)).toMatchSnapshot()
  })

  it('renders correctly when status is NOTE ON', () => {
    shallowExpect(EntryControlSet)(set('entry.status', STATUS_NOTE_ON)(props)).toMatchSnapshot()
  })

  it('renders correctly when status is CC', () => {
    shallowExpect(EntryControlSet)(set('entry.status', STATUS_CONTROL_CHANGE)(props)).toMatchSnapshot()
  })

  it('responds to status change', () => {
    const wrapper = shallow(<EntryControlSet {...props} />)
    wrapper.find('StatusPicker').simulate('change', 45)
    expect(changeStatus).toHaveBeenCalledWith(groupId, isInput, 45)
  })

  it('responds to channel change', () => {
    const wrapper = shallow(<EntryControlSet {...props} />)
    wrapper.find('WithStyles(ChannelPicker)').simulate('change', 11)
    expect(changeChannel).toHaveBeenCalledWith(groupId, isInput, 11)
  })

  it('responds to value change', () => {
    const wrapper = shallow(<EntryControlSet {...props} />)
    wrapper.find('ValuePicker').simulate('change', 3)
    expect(changeValue).toHaveBeenCalledWith(groupId, isInput, 3)
  })

  it('responds to enter being pressed', () => {
    const wrapper = shallow(<EntryControlSet {...props} />)
    wrapper.find('ValuePicker').simulate('keydown', { key: 'Enter', preventDefault: f => f })
    expect(onPressedEnter).toHaveBeenCalledWith(isInput, false, okButtonRef)
  })

  it('responds to tab being pressed', () => {
    const wrapper = shallow(<EntryControlSet {...props} />)
    wrapper.find('ValuePicker').simulate('keydown', { key: 'Tab', preventDefault: f => f })
    expect(onPressedEnter).toHaveBeenCalledWith(isInput, true, okButtonRef)
  })
})
