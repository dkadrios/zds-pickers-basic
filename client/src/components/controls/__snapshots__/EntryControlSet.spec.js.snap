// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryControlSet tests renders correctly as input 1`] = `
<div>
  <h3>
    Input
  </h3>
  <StatusPicker
    channel={9}
    className=""
    mini={false}
    onChange={[Function]}
    status={13}
    statuses={
      Array [
        Object {
          "label": "Note On",
          "value": 9,
        },
        Object {
          "label": "Control Change",
          "value": 11,
        },
        Object {
          "label": "Program Change",
          "value": 12,
        },
        Object {
          "label": "Aftertouch",
          "value": 10,
        },
        Object {
          "label": "Channel Pressure",
          "value": 13,
        },
      ]
    }
    value={22}
  />
  <WithStyles(ChannelPicker)
    channel={9}
    onChange={[Function]}
    status={13}
    value={22}
  />
  <ValuePicker
    autoFocus={true}
    channel={9}
    className=""
    inputRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    status={13}
    value={22}
  />
</div>
`;

exports[`EntryControlSet tests renders correctly as output 1`] = `
<div>
  <h3>
    Output
  </h3>
  <StatusPicker
    channel={9}
    className=""
    mini={false}
    onChange={[Function]}
    status={13}
    statuses={
      Array [
        Object {
          "label": "Note On",
          "value": 9,
        },
        Object {
          "label": "Control Change",
          "value": 11,
        },
        Object {
          "label": "Program Change",
          "value": 12,
        },
        Object {
          "label": "Aftertouch",
          "value": 10,
        },
        Object {
          "label": "Channel Pressure",
          "value": 13,
        },
      ]
    }
    value={22}
  />
  <WithStyles(ChannelPicker)
    channel={9}
    onChange={[Function]}
    status={13}
    value={22}
  />
  <ValuePicker
    autoFocus={false}
    channel={9}
    className=""
    inputRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    status={13}
    value={22}
  />
</div>
`;

exports[`EntryControlSet tests renders correctly when status is CC 1`] = `
<div>
  <h3>
    Output
  </h3>
  <StatusPicker
    channel={9}
    className=""
    mini={false}
    onChange={[Function]}
    status={11}
    statuses={
      Array [
        Object {
          "label": "Note On",
          "value": 9,
        },
        Object {
          "label": "Control Change",
          "value": 11,
        },
        Object {
          "label": "Program Change",
          "value": 12,
        },
        Object {
          "label": "Aftertouch",
          "value": 10,
        },
        Object {
          "label": "Channel Pressure",
          "value": 13,
        },
      ]
    }
    value={22}
  />
  <WithStyles(ChannelPicker)
    channel={9}
    onChange={[Function]}
    status={11}
    value={22}
  />
  <WithStyles(CCPicker)
    autoFocus={false}
    channel={9}
    inputRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    status={11}
    value={22}
  />
</div>
`;

exports[`EntryControlSet tests renders correctly when status is NOTE ON 1`] = `
<div>
  <h3>
    Output
  </h3>
  <StatusPicker
    channel={9}
    className=""
    mini={false}
    onChange={[Function]}
    status={9}
    statuses={
      Array [
        Object {
          "label": "Note On",
          "value": 9,
        },
        Object {
          "label": "Control Change",
          "value": 11,
        },
        Object {
          "label": "Program Change",
          "value": 12,
        },
        Object {
          "label": "Aftertouch",
          "value": 10,
        },
        Object {
          "label": "Channel Pressure",
          "value": 13,
        },
      ]
    }
    value={22}
  />
  <WithStyles(ChannelPicker)
    channel={9}
    onChange={[Function]}
    status={9}
    value={22}
  />
  <NotePicker
    autoFocus={false}
    channel={9}
    disabled={false}
    inputRef={
      Object {
        "current": null,
      }
    }
    mapping={
      Array [
        Object {
          "group": "",
          "name": "",
          "note": 1,
        },
        Object {
          "group": "",
          "name": "",
          "note": 2,
        },
        Object {
          "group": "",
          "name": "",
          "note": 3,
        },
        Object {
          "group": "",
          "name": "",
          "note": 4,
        },
        Object {
          "group": "",
          "name": "",
          "note": 5,
        },
        Object {
          "group": "",
          "name": "",
          "note": 6,
        },
        Object {
          "group": "",
          "name": "",
          "note": 7,
        },
        Object {
          "group": "",
          "name": "",
          "note": 8,
        },
        Object {
          "group": "",
          "name": "",
          "note": 9,
        },
        Object {
          "group": "",
          "name": "",
          "note": 10,
        },
        Object {
          "group": "",
          "name": "",
          "note": 11,
        },
        Object {
          "group": "",
          "name": "",
          "note": 12,
        },
        Object {
          "group": "",
          "name": "",
          "note": 13,
        },
        Object {
          "group": "",
          "name": "",
          "note": 14,
        },
        Object {
          "group": "",
          "name": "",
          "note": 15,
        },
        Object {
          "group": "",
          "name": "",
          "note": 16,
        },
        Object {
          "group": "",
          "name": "",
          "note": 17,
        },
        Object {
          "group": "",
          "name": "",
          "note": 18,
        },
        Object {
          "group": "",
          "name": "",
          "note": 19,
        },
        Object {
          "group": "",
          "name": "",
          "note": 20,
        },
        Object {
          "group": "",
          "name": "",
          "note": 21,
        },
        Object {
          "group": "",
          "name": "",
          "note": 22,
        },
        Object {
          "group": "",
          "name": "",
          "note": 23,
        },
        Object {
          "group": "Kicks",
          "name": "Kick",
          "note": 24,
        },
        Object {
          "group": "Snares",
          "name": "Side Stick",
          "note": 25,
        },
        Object {
          "group": "Snares",
          "name": "Snare",
          "note": 26,
        },
        Object {
          "group": "Snares",
          "name": "Snare Drag",
          "note": 27,
        },
        Object {
          "group": "Snares",
          "name": "Snare Half Edge",
          "note": 28,
        },
        Object {
          "group": "Snares",
          "name": "Snare Rimshot",
          "note": 29,
        },
        Object {
          "group": "Hats",
          "name": "HH Closed Tip",
          "note": 30,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 1",
          "note": 31,
        },
        Object {
          "group": "Hats",
          "name": "HH Pedal",
          "note": 32,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 2",
          "note": 33,
        },
        Object {
          "group": "Hats",
          "name": "HH Open Tip",
          "note": 34,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom 1",
          "note": 35,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom 2",
          "note": 36,
        },
        Object {
          "group": "Hats",
          "name": "HH Closed Shank",
          "note": 37,
        },
        Object {
          "group": "Toms",
          "name": "High Tom 1",
          "note": 38,
        },
        Object {
          "group": "Hats",
          "name": "HH Open Shank",
          "note": 39,
        },
        Object {
          "group": "Toms",
          "name": "High Tom 2",
          "note": 40,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 1 Edge",
          "note": 41,
        },
        Object {
          "group": "Hats",
          "name": "HH 1/4 Open Tip",
          "note": 42,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 2 Edge",
          "note": 43,
        },
        Object {
          "group": "Hats",
          "name": "HH 1/4 Open Shank",
          "note": 44,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 1 Edge",
          "note": 45,
        },
        Object {
          "group": "Hats",
          "name": "HH 1/2 Open Tip",
          "note": 46,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 2 Edge",
          "note": 47,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 3 Edge",
          "note": 48,
        },
        Object {
          "group": "Hats",
          "name": "HH 1/2 Open Shank",
          "note": 49,
        },
        Object {
          "group": "Rides",
          "name": "Ride 1 Edge",
          "note": 50,
        },
        Object {
          "group": "Hats",
          "name": "HH 3/4 Open Tip",
          "note": 51,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 1 Bow",
          "note": 52,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 2 Bow",
          "note": 53,
        },
        Object {
          "group": "Hats",
          "name": "HH 3/4 Open Shank",
          "note": 54,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 1 Bow",
          "note": 55,
        },
        Object {
          "group": "Hats",
          "name": "HH Splash",
          "note": 56,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 2 Bow",
          "note": 57,
        },
        Object {
          "group": "Hats",
          "name": "HH Bell Tip",
          "note": 58,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 3 Bow",
          "note": 59,
        },
        Object {
          "group": "Rides",
          "name": "Ride 1 Bow",
          "note": 60,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 1 Rimshot",
          "note": 61,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 1 Bell",
          "note": 62,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 2 Rimshot",
          "note": 63,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 2 Bell",
          "note": 64,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 1 Bell",
          "note": 65,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom 1 Rimshot",
          "note": 66,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 2 Bell",
          "note": 67,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom 2 Rimshot",
          "note": 68,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 3 Bell",
          "note": 69,
        },
        Object {
          "group": "Toms",
          "name": "High Tom 1 Rimshot",
          "note": 70,
        },
        Object {
          "group": "Rides",
          "name": "Ride 1 Bell",
          "note": 71,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 1 Hit",
          "note": 72,
        },
        Object {
          "group": "Toms",
          "name": "High Tom 2 Rimshot",
          "note": 73,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 2 Hit",
          "note": 74,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 1 Choke",
          "note": 75,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 1 Hit",
          "note": 76,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 2 Hit",
          "note": 77,
        },
        Object {
          "group": "Cymbals",
          "name": "Crash 2 Choke",
          "note": 78,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 3 Hit",
          "note": 79,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 1 Choke",
          "note": 80,
        },
        Object {
          "group": "Rides",
          "name": "Ride 1 Hit",
          "note": 81,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 2 Choke",
          "note": 82,
        },
        Object {
          "group": "Perc",
          "name": "Perc Hit",
          "note": 83,
        },
        Object {
          "group": "Kicks",
          "name": "Kick No Snare",
          "note": 84,
        },
        Object {
          "group": "Cymbals",
          "name": "Cymbal 3 Choke",
          "note": 85,
        },
        Object {
          "group": "Snares",
          "name": "Snare Flam",
          "note": 86,
        },
        Object {
          "group": "Rides",
          "name": "Ride 1 Choke",
          "note": 87,
        },
        Object {
          "group": "Snares",
          "name": "Snare Rim Click",
          "note": 88,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 1 Rim Click",
          "note": 89,
        },
        Object {
          "group": "Toms",
          "name": "Floor Tom 2 Rim Click",
          "note": 90,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom Rim 1 Click",
          "note": 91,
        },
        Object {
          "group": "Toms",
          "name": "Mid Tom 2 Rim Click",
          "note": 92,
        },
        Object {
          "group": "Toms",
          "name": "High Tom Rim 1 Click",
          "note": 93,
        },
        Object {
          "group": "Toms",
          "name": "High Tom 2 Rim Click",
          "note": 94,
        },
        Object {
          "group": "Perc",
          "name": "Perc 2 Hit",
          "note": 95,
        },
        Object {
          "group": "Groove",
          "name": "Groove 1",
          "note": 96,
        },
        Object {
          "group": "Groove",
          "name": "Groove 2",
          "note": 97,
        },
        Object {
          "group": "Groove",
          "name": "Groove 3",
          "note": 98,
        },
        Object {
          "group": "Groove",
          "name": "Groove 4",
          "note": 99,
        },
        Object {
          "group": "Groove",
          "name": "Groove 5",
          "note": 100,
        },
        Object {
          "group": "Groove",
          "name": "Groove 6",
          "note": 101,
        },
        Object {
          "group": "Groove",
          "name": "Groove 7",
          "note": 102,
        },
        Object {
          "group": "Groove",
          "name": "Groove 8",
          "note": 103,
        },
        Object {
          "group": "Groove",
          "name": "Groove 9",
          "note": 104,
        },
        Object {
          "group": "Groove",
          "name": "Groove 10",
          "note": 105,
        },
        Object {
          "group": "Groove",
          "name": "Groove 11",
          "note": 106,
        },
        Object {
          "group": "Groove",
          "name": "Groove 12",
          "note": 107,
        },
        Object {
          "group": "Groove",
          "name": "Groove 13",
          "note": 108,
        },
        Object {
          "group": "Groove",
          "name": "Groove 14",
          "note": 109,
        },
        Object {
          "group": "Groove",
          "name": "Groove 15",
          "note": 110,
        },
        Object {
          "group": "Groove",
          "name": "Groove 16",
          "note": 111,
        },
        Object {
          "group": "Groove",
          "name": "Groove 17",
          "note": 112,
        },
        Object {
          "group": "Groove",
          "name": "Groove 18",
          "note": 113,
        },
        Object {
          "group": "Groove",
          "name": "Groove 19",
          "note": 114,
        },
        Object {
          "group": "Groove",
          "name": "Groove 20",
          "note": 115,
        },
        Object {
          "group": "Groove",
          "name": "Groove 21",
          "note": 116,
        },
        Object {
          "group": "Groove",
          "name": "Groove 22",
          "note": 117,
        },
        Object {
          "group": "Groove",
          "name": "Groove 23",
          "note": 118,
        },
        Object {
          "group": "Groove",
          "name": "Groove 24",
          "note": 119,
        },
        Object {
          "group": "Groove",
          "name": "Groove 25",
          "note": 120,
        },
        Object {
          "group": "Groove",
          "name": "Groove 26",
          "note": 121,
        },
        Object {
          "group": "Groove",
          "name": "Groove 27",
          "note": 122,
        },
        Object {
          "group": "Groove",
          "name": "Groove 28",
          "note": 123,
        },
        Object {
          "group": "Groove",
          "name": "Groove 29",
          "note": 124,
        },
        Object {
          "group": "Groove",
          "name": "Groove 30",
          "note": 125,
        },
        Object {
          "group": "Groove",
          "name": "Groove 31",
          "note": 126,
        },
        Object {
          "group": "Groove",
          "name": "Groove 32",
          "note": 127,
        },
        Object {
          "group": "",
          "name": "",
          "note": 128,
        },
      ]
    }
    onChange={[Function]}
    onKeyDown={[Function]}
    status={9}
    value={22}
  />
</div>
`;
